File: packages/python/tests/utils/get_extension_test.py
import pytest
from pathlib import Path
from cachetta.utils.get_extension import get_extension
from cachetta.exceptions import InvalidPathError

def describe_get_extension():
    def test_it_works_with_a_str():
        assert get_extension('foo/bar/baz.pkl') == 'pkl'

    def test_it_works_with_a_path():
        assert get_extension(Path('foo/bar/baz.pkl')) == 'pkl'

    def test_it_works_with_multiple_dots():
        assert get_extension('foo.bar.baz.pkl') == 'pkl'

    def test_it_raises_if_missing_extension_with_str():
        with pytest.raises(InvalidPathError):
            get_extension('foo/bar/baz')

    def test_it_raises_if_missing_extension_with_path():
        with pytest.raises(InvalidPathError):
            get_extension(Path('foo/bar/baz'))

    def test_it_returns_extension_of_simple_file():
        assert get_extension('test.txt') == 'txt'

    def test_it_returns_extension_of_file_with_multiple_dots():
        assert get_extension('test.min.js') == 'js'
        assert get_extension('config.prod.json') == 'json'

    def test_it_returns_extension_of_file_with_path():
        assert get_extension('/path/to/file.json') == 'json'
        assert get_extension('./cache/data.pkl') == 'pkl'
        assert get_extension('../config/settings.yaml') == 'yaml'

    def test_it_returns_extension_of_file_with_complex_path():
        assert get_extension('/home/user/projects/cache/data.json') == 'json'
        assert get_extension('C:\\Users\\username\\Documents\\file.txt') == 'txt'

    def test_it_handles_different_file_extensions():
        assert get_extension('data.json') == 'json'
        assert get_extension('data.pkl') == 'pkl'
        assert get_extension('data.yaml') == 'yaml'
        assert get_extension('data.yml') == 'yml'
        assert get_extension('data.xml') == 'xml'
        assert get_extension('data.csv') == 'csv'



    def test_it_raises_error_for_empty_string():
        with pytest.raises(InvalidPathError):
            get_extension('')

File: packages/javascript/src/utils/get-extension.test.ts
import { describe, test, expect } from 'vitest';
import { getExtension } from './get-extension.js';
import { CachettaError } from '../errors.js';

describe('getExtension', () => {
  test('should return the extension of a simple file', () => {
    expect(getExtension('test.txt')).toBe('txt');
  });

  test('should return the extension of a file with multiple dots', () => {
    expect(getExtension('test.min.js')).toBe('js');
    expect(getExtension('config.prod.json')).toBe('json');
  });

  test('should return the extension of a file with path', () => {
    expect(getExtension('/path/to/file.json')).toBe('json');
    expect(getExtension('./cache/data.pkl')).toBe('pkl');
    expect(getExtension('../config/settings.yaml')).toBe('yaml');
  });

  test('should return the extension of a file with complex path', () => {
    expect(getExtension('/home/user/projects/cache/data.json')).toBe('json');
    expect(getExtension('C:\\Users\\username\\Documents\\file.txt')).toBe('txt');
  });

  test('should handle different file extensions', () => {
    expect(getExtension('data.json')).toBe('json');
    expect(getExtension('data.pkl')).toBe('pkl');
    expect(getExtension('data.yaml')).toBe('yaml');
    expect(getExtension('data.yml')).toBe('yml');
    expect(getExtension('data.xml')).toBe('xml');
    expect(getExtension('data.csv')).toBe('csv');
  });

  test('should throw CachettaError for file without extension', () => {
    expect(() => getExtension('file')).toThrow(CachettaError);
    expect(() => getExtension('/path/to/file')).toThrow(CachettaError);
    expect(() => getExtension('./cache/data')).toThrow(CachettaError);
  });

  test('should throw CachettaError for file ending with dot', () => {
    expect(() => getExtension('file.')).toThrow(CachettaError);
    expect(() => getExtension('/path/to/file.')).toThrow(CachettaError);
  });

  test('should handle PathLike objects', () => {
    const pathLike = new URL('file:///test.json');
    expect(getExtension(pathLike)).toBe('json');
  });

  test('should handle empty string', () => {
    expect(() => getExtension('')).toThrow(CachettaError);
  });
});
